### ISSUE TITLE
Allow more than 128MB of Z3 RAM

### ISSUE THREAD

**Issue by DesertBlizzard** [midwan/amiberry#390]
Option to assign more than 128MB fast memory.

**Comment by CypherXG** [comment]
128 megabytes are really very much for an Amiga. Which program needs so much memory? The other way around, one must not forget that a Pi has only 1024MB from which the buffer for the graphics card is subtracted, the operating system and the programs ...
Write in the terminal free ....
Then you see what you still from the RAM of the Raspberry available

**Comment by midwan** [comment]
@DesertBlizzard 
Do you have a real-life scenario that requires more RAM?
I know that WinUAE allows for allocating more, but Amiberry is targetting less powerful ARM boards which usually come with much less RAM also. As mentioned above, a Raspberry Pi comes with 1GB of RAM (or less, if you look at the new low-cost models!) and that is shared between the system and the GPU.

You can already allocate up to 128MB of Z3 Fast RAM, another 128MB of RTG Ram, 8MB of Chip RAM. I haven't seen a scenario that requires more under emulation, unless you're running things like php, MySQL and web servers at the same time.

**Comment by DesertBlizzard** [comment]
I actually have an extra gig laying waste on the tinkerboard.  I guess to answer your question, 1) Gates once said we'd never need more that 640KB, and 2) because we can.

I'd especially like to see the WinUAE option for 32bit Chip Ram come to Amiberry.  Perhaps the ability to load up to 512MB of fast ram would be nice.  High-Res true color images take up a lot of RAM when you manipulate them.  So my thought is if the resources are there and possible to implement, why not?  Give users the tools to be creative.

**Comment by CypherXG** [comment]
Are you using a Amiga Emulation or are you using your High End PC to edit "High-Res true color images" ?
I'm sure that you are using PS and not your Amiga for this. Also, a really Big High-Res true color image never use 512MB and you are not edit pictures in raw mode...
I have some High-Res true color images with 7000x11000px, and they need ~450MB Ram...but you can never edit these size on an Amiga...
You cannot compare an arm computer with an High End PC. Sure, if you have 16GB Ram you can power up Winuae with 1GB 32bit Chip Ram and 1GB Fast....but which program use so many memory on the Amiga or can even recognize this memory size?
Same with old programs. They can also not regonize HD images bigger than 4GB....

**Comment by DesertBlizzard** [comment]
![20181230_073007158_ios](https://user-images.githubusercontent.com/45675403/50545233-5fcbc580-0bca-11e9-9f8e-ca2f94861f6c.jpg)
Nearly bone stock install of Amikit X.  Netsurf 3.8 open, the only application with the page show loaded.  Follow the download link and begin downloading the 5MB demo.
![20181230_073453335_ios](https://user-images.githubusercontent.com/45675403/50545249-f4362800-0bca-11e9-8f29-8f4fc38fe41f.jpg)
Extract the LHA to RAM:
![20181230_073806946_ios](https://user-images.githubusercontent.com/45675403/50545261-5131de00-0bcb-11e9-9fa8-071e62d935f2.jpg)
Free memory at this point is only 16.4MB of the 103MB that was available at the start of this one simple task.
![20181230_074135930_ios](https://user-images.githubusercontent.com/45675403/50545278-cc938f80-0bcb-11e9-9deb-2b41a9eab6de.jpg)
Plenty of unused memory on the host.
![20181230_074657458_ios](https://user-images.githubusercontent.com/45675403/50545326-d2d63b80-0bcc-11e9-9d69-8c3e7d4d46d2.jpg)

This is one real-world example of a need for more memory to be available.  This tinekerboard is pretty capable unit!

Sorry for the upside-down image.

**Comment by midwan** [comment]
@DesertBlizzard 
Thanks for the feedback. It would be nice to have more RAM under emulation for a future version, but it's more work than just a slider I'm afraid. The memory needs to be allocated safely, with fallbacks if something goes wrong, and we need to manage all the different types of memory also (Chip, Fast, Z3, RTG, Mainboard RAM). With the current implementation of the code, which came from Uae4Arm, if we increase the amount of RAM that we can allocate for Z3 RAM, the whole emulator will crash because it wasn't designed for that.

In other words, we'll need to rewrite the memory allocation part at least partially, to allow for more RAM within the emulator. It might happen at some point, but I don't see it as a high priority issue right now.

**Comment by DesertBlizzard** [comment]
I understand.  I would hope to see this simply added to the list of enhancements. Certainly, I can work with 128MB of RAM, butnto have more never hurts and helps me realize the potential of my HW. 

On a side note, is AHI as complicated to implement?  😁😁😁

**Comment by HoraceAndTheSpider** [comment]
At no point did I call anyone “stupid” - however I think it has be to be accepted this is not a typical requirement, and was already well explained by others. 

However the “just a slider” comment implied a lack of understanding about the complexity of the implementation.

**Comment by DesertBlizzard** [comment]
In all honesty, I don’t have a full understanding of the complexity involved. What I do understand is that the code to do this exists on other platforms, so from the non-developer point of view it would seem that the wheel doesn’t necessarily have to be reinvented. Nonetheless, I wish I had the chops to directly help in this regard. 

What is a typical requirement of this software?  Why provide bsdsocket, RTG, 8MB Chip RAM, Z3 Fast RAM, 040 emulation, etc., if the intent of the software is to only run play old OCS/ECS/AGA games?  Is that the scope of the goal for this project?  My intent for the software would be to obtain the full extent of a high-end Amiga experience on my amazing little SBC.  I would like to use AminetRadio, but cannot until AHI support is enabled. Is that beyond the intent? I believe others would also like to push the envelope, but perhaps I’m an outlier pushing the boundaries of the software.  I do hope more people would do the same though. 

I have immense gratitude that this software even exists. Many thanks to all contributors.

**Comment by CypherXG** [comment]
I see u are using AmikitX....
How many memory is free after a fresh start because i think with all that tools which are running in the backround you loose a lot of memory.
Also you must think that amiberry is written to use it with Winuae where you can have Gigs of memory and not for small arm machines.

**Comment by midwan** [comment]
Now that we have new RPI models out, with 4GB RAM available, we should revisit this issue. :)

**Comment by CypherXG** [comment]
AHI support.....Yehaaaa....that would be nice....

**Comment by midwan** [comment]
Related: 9deeed16a2fe147dc77b9e5b067b259469f83cad

**Comment by midwan** [comment]
You can now have a combined maximum of 768 MB of Fast RAM (512MB Z3, 128MB Processor slot).
